# <div align="center"> 山东英才 </div>

In [2]:
import os
import json
import xlrd
import hashlib
import pandas as pd
from k12libs.k12ai.k12ai_errmsg import ERRORS
from IPython.display import display, Markdown, HTML

In [40]:
display(HTML('<style>.container { width:%d%% !important; }</style>' % 80))

## 问题单

In [5]:
students_all = ['10110207020%02d' % (i+1) for i in range(40)]

### 20210107: 10:00  ~ 11:00

#### 在线人数

In [6]:
students_online = [
    '1011020702001', 
    '1011020702003', 
    '1011020702004', 
    '1011020702005', 
    '1011020702006', 
    '1011020702007', 
    '1011020702008', 
    '1011020702009', 
    '1011020702012', 
    '1011020702013', 
    '1011020702014', 
    '1011020702015', 
    '1011020702016', 
    '1011020702018', 
    '1011020702019', 
    '1011020702020', 
    '1011020702021', 
    '1011020702022', 
    '1011020702023', 
    '1011020702025', 
    '1011020702026', 
    '1011020702027', 
    '1011020702028', 
    '1011020702029', 
    '1011020702030', 
]

print(len(students_online), students_online)

25 ['1011020702001', '1011020702003', '1011020702004', '1011020702005', '1011020702006', '1011020702007', '1011020702008', '1011020702009', '1011020702012', '1011020702013', '1011020702014', '1011020702015', '1011020702016', '1011020702018', '1011020702019', '1011020702020', '1011020702021', '1011020702022', '1011020702023', '1011020702025', '1011020702026', '1011020702027', '1011020702028', '1011020702029', '1011020702030']


#### 离线线人数

In [7]:
students_offline = [x for x in students_all if x not in students_online]
print(len(students_offline), students_offline)

15 ['1011020702002', '1011020702010', '1011020702011', '1011020702017', '1011020702024', '1011020702031', '1011020702032', '1011020702033', '1011020702034', '1011020702035', '1011020702036', '1011020702037', '1011020702038', '1011020702039', '1011020702040']


#### 错误统计

In [19]:
def xls_parse_errors(xls_file):
    book = xlrd.open_workbook(xls_file)
    sheet = book.sheet_by_index(0)
    errors = {
        'unkown': {},
        'general': {}
    }

    for rowid in range(1, sheet.nrows):
        user, uuid, errcode, payload = sheet.row_values(rowid)
        errcode = int(errcode)
        payload = json.loads(payload)
        if errcode == 999999:
            text = payload['errinfo']['err_text']
            md5 = hashlib.md5(text.encode()).hexdigest()[0:6]
            if md5 not in errors['unkown']:
                errors['unkown'][md5] = {
                    'err_text': text,
                    'trackback': json.dumps(payload['errinfo']['trackback'], indent=2).replace('\n', '</br>')
                }
        elif errcode > 100100:
            if errcode not in errors['general'] and 'trackback' in payload['errinfo']:
                errors['general'][errcode] = {
                    'err_text': payload['errinfo']['err_text'],
                    'trackback': json.dumps(payload['errinfo']['trackback'], indent=2).replace('\n', '</br>')
                }
    return errors

---------------------

##### General错误

In [27]:
xls_file = './sdyc/20210107报错信息.xls'
errors = xls_parse_errors(xls_file)

In [23]:
def general_error_md(errors):
    data = '''ErrorCode | ErrorText | Trackback
    :---:|:---|:---
    <img width=150/> | <img width=250/> | <img width=350/>
    '''

    for errcode, errinfo in errors['general'].items():
        data += f'{errcode}|{errinfo["err_text"]}|{errinfo["trackback"]}\n'
    return data

Markdown(general_error_md(errors))

ErrorCode | ErrorText | Trackback
    :---:|:---|:---
    <img width=150/> | <img width=250/> | <img width=350/>
    100404|name 'Input_oth1flre4i_feat' is not defined|[</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/lib/runner/controller.py",</br>    "linenum": 37,</br>    "funcname": "train",</br>    "source": "runner.train()"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/runner/cls/image_classifier.py",</br>    "linenum": 121,</br>    "funcname": "train",</br>    "source": "out = self.cls_net(data_dict['img'])"</br>  },</br>  {</br>    "filename": "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py",</br>    "linenum": 550,</br>    "funcname": "__call__",</br>    "source": "result = self.forward(*input, **kwargs)"</br>  },</br>  {</br>    "filename": "/usr/local/lib/python3.6/dist-packages/torch/nn/parallel/data_parallel.py",</br>    "linenum": 150,</br>    "funcname": "forward",</br>    "source": "return self.module(*inputs[0], **kwargs[0])"</br>  },</br>  {</br>    "filename": "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py",</br>    "linenum": 550,</br>    "funcname": "__call__",</br>    "source": "result = self.forward(*input, **kwargs)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/model/cls/nets/base_model.py",</br>    "linenum": 61,</br>    "funcname": "forward",</br>    "source": "return self.net(inputs)"</br>  },</br>  {</br>    "filename": "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py",</br>    "linenum": 550,</br>    "funcname": "__call__",</br>    "source": "result = self.forward(*input, **kwargs)"</br>  },</br>  {</br>    "filename": "/cache/output/custom/custom_base.py",</br>    "linenum": 19,</br>    "funcname": "forward",</br>    "source": "Flatten_m7okrjm4d_feat = self.Flatten_m7okrjm4d(Input_oth1flre4i_feat)"</br>  }</br>]
100401|size mismatch, m1: [21504 x 224], m2: [10 x 22] at /pytorch/aten/src/THC/generic/THCTensorMathBlas.cu:283|[</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/lib/runner/controller.py",</br>    "linenum": 37,</br>    "funcname": "train",</br>    "source": "runner.train()"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/runner/cls/image_classifier.py",</br>    "linenum": 121,</br>    "funcname": "train",</br>    "source": "out = self.cls_net(data_dict['img'])"</br>  },</br>  {</br>    "filename": "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py",</br>    "linenum": 550,</br>    "funcname": "__call__",</br>    "source": "result = self.forward(*input, **kwargs)"</br>  },</br>  {</br>    "filename": "/usr/local/lib/python3.6/dist-packages/torch/nn/parallel/data_parallel.py",</br>    "linenum": 150,</br>    "funcname": "forward",</br>    "source": "return self.module(*inputs[0], **kwargs[0])"</br>  },</br>  {</br>    "filename": "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py",</br>    "linenum": 550,</br>    "funcname": "__call__",</br>    "source": "result = self.forward(*input, **kwargs)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/model/cls/nets/base_model.py",</br>    "linenum": 61,</br>    "funcname": "forward",</br>    "source": "return self.net(inputs)"</br>  },</br>  {</br>    "filename": "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py",</br>    "linenum": 550,</br>    "funcname": "__call__",</br>    "source": "result = self.forward(*input, **kwargs)"</br>  },</br>  {</br>    "filename": "/cache/output/custom/custom_base.py",</br>    "linenum": 18,</br>    "funcname": "forward",</br>    "source": "Linear_uhzliwikfn_feat = self.Linear_uhzliwikfn(x)"</br>  },</br>  {</br>    "filename": "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py",</br>    "linenum": 550,</br>    "funcname": "__call__",</br>    "source": "result = self.forward(*input, **kwargs)"</br>  },</br>  {</br>    "filename": "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/linear.py",</br>    "linenum": 87,</br>    "funcname": "forward",</br>    "source": "return F.linear(input, self.weight, self.bias)"</br>  },</br>  {</br>    "filename": "/usr/local/lib/python3.6/dist-packages/torch/nn/functional.py",</br>    "linenum": 1612,</br>    "funcname": "linear",</br>    "source": "output = input.matmul(weight.t())"</br>  }</br>]
100403|__init__() missing 1 required positional argument: 'in_channels'|[</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/runner/runner_selector.py",</br>    "linenum": 114,</br>    "funcname": "cls_runner",</br>    "source": "return CLS_METHOD_DICT[key](self.configer)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/runner/cls/image_classifier.py",</br>    "linenum": 39,</br>    "funcname": "__init__",</br>    "source": "self.cls_net = self.cls_model_manager.get_cls_model()"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/model/cls/model_manager.py",</br>    "linenum": 33,</br>    "funcname": "get_cls_model",</br>    "source": "model = CLS_MODEL_DICT[model_name](self.configer)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/app/k12ai/model/cls/nets/custom_base.py",</br>    "linenum": 21,</br>    "funcname": "__init__",</br>    "source": "self.net = load_custom_model(cache_dir, model_name)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/app/k12ai/tools/util/net_def.py",</br>    "linenum": 32,</br>    "funcname": "load_custom_model",</br>    "source": "custom_model = custom_net.custom_model()"</br>  },</br>  {</br>    "filename": "/cache/output/custom/custom_base.py",</br>    "linenum": 29,</br>    "funcname": "custom_model",</br>    "source": "return Hproject_model_name()"</br>  },</br>  {</br>    "filename": "/cache/output/custom/custom_base.py",</br>    "linenum": 11,</br>    "funcname": "__init__",</br>    "source": "self.Conv2d_ptvet7fj4 = nn.Conv2d(out_channels=64, kernel_size=3, stride=1, padding=1, dilation=1, groups=1, bias=True)"</br>  }</br>]


---------------------

##### Unkown错误

In [24]:
def unkown_error_md(errors):
    data = '''ErrorCode | ErrorText | Trackback
    :---:|:---|:---
    <img width=150/> | <img width=250/> | <img width=350/>
    '''

    for errcode, errinfo in errors['unkown'].items():
        data += f'{errcode}|{errinfo["err_text"]}|{errinfo["trackback"]}\n'
    return data

Markdown(unkown_error_md(errors))

ErrorCode | ErrorText | Trackback
    :---:|:---|:---
    <img width=150/> | <img width=250/> | <img width=350/>
    4788bf|__init__() missing 1 required positional argument: 'in_features'|[</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/runner/runner_selector.py",</br>    "linenum": 114,</br>    "funcname": "cls_runner",</br>    "source": "return CLS_METHOD_DICT[key](self.configer)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/runner/cls/image_classifier.py",</br>    "linenum": 39,</br>    "funcname": "__init__",</br>    "source": "self.cls_net = self.cls_model_manager.get_cls_model()"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/model/cls/model_manager.py",</br>    "linenum": 33,</br>    "funcname": "get_cls_model",</br>    "source": "model = CLS_MODEL_DICT[model_name](self.configer)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/app/k12ai/model/cls/nets/custom_base.py",</br>    "linenum": 21,</br>    "funcname": "__init__",</br>    "source": "self.net = load_custom_model(cache_dir, model_name)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/app/k12ai/tools/util/net_def.py",</br>    "linenum": 32,</br>    "funcname": "load_custom_model",</br>    "source": "custom_model = custom_net.custom_model()"</br>  },</br>  {</br>    "filename": "/cache/output/custom/custom_base.py",</br>    "linenum": 27,</br>    "funcname": "custom_model",</br>    "source": "return Hproject_model_name()"</br>  },</br>  {</br>    "filename": "/cache/output/custom/custom_base.py",</br>    "linenum": 15,</br>    "funcname": "__init__",</br>    "source": "self.Linear_x1sub019a = nn.Linear(out_features=22, bias=True)"</br>  }</br>]
173d43|Trying to create tensor with negative dimension -392: [10, -392]|[</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/runner/runner_selector.py",</br>    "linenum": 114,</br>    "funcname": "cls_runner",</br>    "source": "return CLS_METHOD_DICT[key](self.configer)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/runner/cls/image_classifier.py",</br>    "linenum": 39,</br>    "funcname": "__init__",</br>    "source": "self.cls_net = self.cls_model_manager.get_cls_model()"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/model/cls/model_manager.py",</br>    "linenum": 33,</br>    "funcname": "get_cls_model",</br>    "source": "model = CLS_MODEL_DICT[model_name](self.configer)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/app/k12ai/model/cls/nets/custom_base.py",</br>    "linenum": 21,</br>    "funcname": "__init__",</br>    "source": "self.net = load_custom_model(cache_dir, model_name)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/app/k12ai/tools/util/net_def.py",</br>    "linenum": 32,</br>    "funcname": "load_custom_model",</br>    "source": "custom_model = custom_net.custom_model()"</br>  },</br>  {</br>    "filename": "/cache/output/custom/custom_base.py",</br>    "linenum": 37,</br>    "funcname": "custom_model",</br>    "source": "return Hproject_model_name()"</br>  },</br>  {</br>    "filename": "/cache/output/custom/custom_base.py",</br>    "linenum": 19,</br>    "funcname": "__init__",</br>    "source": "self.Linear_dmz8bluxar = nn.Linear(in_features=-392, out_features=10, bias=True)"</br>  },</br>  {</br>    "filename": "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/linear.py",</br>    "linenum": 72,</br>    "funcname": "__init__",</br>    "source": "self.weight = Parameter(torch.Tensor(out_features, in_features))"</br>  }</br>]
dbe4e4|Trying to create tensor with negative dimension -72: [10, -72]|[</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/runner/runner_selector.py",</br>    "linenum": 114,</br>    "funcname": "cls_runner",</br>    "source": "return CLS_METHOD_DICT[key](self.configer)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/runner/cls/image_classifier.py",</br>    "linenum": 39,</br>    "funcname": "__init__",</br>    "source": "self.cls_net = self.cls_model_manager.get_cls_model()"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/model/cls/model_manager.py",</br>    "linenum": 33,</br>    "funcname": "get_cls_model",</br>    "source": "model = CLS_MODEL_DICT[model_name](self.configer)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/app/k12ai/model/cls/nets/custom_base.py",</br>    "linenum": 21,</br>    "funcname": "__init__",</br>    "source": "self.net = load_custom_model(cache_dir, model_name)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/app/k12ai/tools/util/net_def.py",</br>    "linenum": 32,</br>    "funcname": "load_custom_model",</br>    "source": "custom_model = custom_net.custom_model()"</br>  },</br>  {</br>    "filename": "/cache/output/custom/custom_base.py",</br>    "linenum": 37,</br>    "funcname": "custom_model",</br>    "source": "return Hproject_model_name()"</br>  },</br>  {</br>    "filename": "/cache/output/custom/custom_base.py",</br>    "linenum": 19,</br>    "funcname": "__init__",</br>    "source": "self.Linear_dmz8bluxar = nn.Linear(in_features=-72, out_features=10, bias=True)"</br>  },</br>  {</br>    "filename": "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/linear.py",</br>    "linenum": 72,</br>    "funcname": "__init__",</br>    "source": "self.weight = Parameter(torch.Tensor(out_features, in_features))"</br>  }</br>]
723c6b|__init__() missing 1 required positional argument: 'num_features'|[</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/runner/runner_selector.py",</br>    "linenum": 114,</br>    "funcname": "cls_runner",</br>    "source": "return CLS_METHOD_DICT[key](self.configer)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/runner/cls/image_classifier.py",</br>    "linenum": 39,</br>    "funcname": "__init__",</br>    "source": "self.cls_net = self.cls_model_manager.get_cls_model()"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/model/cls/model_manager.py",</br>    "linenum": 33,</br>    "funcname": "get_cls_model",</br>    "source": "model = CLS_MODEL_DICT[model_name](self.configer)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/app/k12ai/model/cls/nets/custom_base.py",</br>    "linenum": 21,</br>    "funcname": "__init__",</br>    "source": "self.net = load_custom_model(cache_dir, model_name)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/app/k12ai/tools/util/net_def.py",</br>    "linenum": 32,</br>    "funcname": "load_custom_model",</br>    "source": "custom_model = custom_net.custom_model()"</br>  },</br>  {</br>    "filename": "/cache/output/custom/custom_base.py",</br>    "linenum": 33,</br>    "funcname": "custom_model",</br>    "source": "return Hproject_model_name()"</br>  },</br>  {</br>    "filename": "/cache/output/custom/custom_base.py",</br>    "linenum": 15,</br>    "funcname": "__init__",</br>    "source": "self.BatchNorm2d_f8s67obtuj = nn.BatchNorm2d(momentum=0.1, affine=True, track_running_stats=True)"</br>  }</br>]


#### 用户错误分类

In [13]:
df = pd.read_excel(xls_file, header=0, usecols='A:C')
df_sorted = df[df.error_code > 100100].set_index(['error_code', 'user_code']).sort_index()
df_sorted.index = df_sorted.index.map(lambda x: (f'{x[0]}:{ERRORS[x[0]]["cn"]}', x[1]))
df_sorted

mission_id
error_code                         user_code                
100401:自定义模型错误: 模型上下层size大小不匹配     1011020702019           3
                                   1011020702019           3
                                   1011020702019           3
                                   1011020702026           3
                                   1011020702026           3
100402:自定义模型错误: Pad参数要大于kernel值的一半 1011020702019           3
100403:自定义模型错误: 卷积层缺少in_channels参数 1011020702006           4
                                   1011020702006           4
                                   1011020702006           4
                                   1011020702006           4
                                   1011020702006           4
999999:常见错误: 不可知错误                 1011020702028           3
                                   1011020702028           3
                                   1011020702028           3
                                   1011020702028           3
                                   1011020702028           3
                                   1011020702028           3
                                   1011020702028           3
                                   1011020702028           3
                                   1011020702028           3
                                   1011020702028           3
                                   1011020702028           3
                                   1011020702028           3
                                   1011020702028           3
                                   1011020702028           3
                                   1011020702028           3
                                   1011020702028           3
                                   1011020702028           3
                                   1011020702028           3
                                   1011020702028           3
                                   1011020702028           3
                                   1011020702028           3

#### 错误定位

1. [x] <font color='Green'><strong>中级别</strong></font>100403:自定义模型错误: 卷积层缺少in_channels参数
 > 李东: 只有一个学生(1011020702006)在第4节课出现, 导致原因是Conv层的in_channels输入为空, 该值设计上是不能让用户修改的, 即:绝不可能出现这种情况(除非前端有bug), 建议处理方式-后台手动修改这个层的数据, 纠正过来 ![](./sdyc/err_02.png)
- [ ] <font color='Gray'><strong>低级别</strong></font>100402:自定义模型错误: Pad参数要大于kernel值的一半
 > 李东: 模型已经被修改 (1011020702019 不能追溯)
- [ ] <font color='Gray'><strong>低级别</strong></font>100401:自定义模型错误: 模型上下层size大小不匹配
 > 李东: 模型拖拽错误 (不能追溯)
- [ ] <font color='Gray'><strong>低级别</strong></font>999999:常见错误: 不可知错误
 > 李东: 模型拖拽错误 (1011020702028 不能追溯)

### 20210107: 10:00 ~ 11:00

#### 错误统计

In [59]:
xls_file = 'sdyc/errorinfo_20210114.xls'
errors = xls_parse_errors(xls_file)

##### General错误

In [60]:
Markdown(general_error_md(errors))

ErrorCode | ErrorText | Trackback
    :---:|:---|:---
    <img width=150/> | <img width=250/> | <img width=350/>
    100404|name 'Input_oth1flre4i_feat' is not defined|[</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/lib/runner/controller.py",</br>    "linenum": 37,</br>    "funcname": "train",</br>    "source": "runner.train()"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/runner/cls/image_classifier.py",</br>    "linenum": 121,</br>    "funcname": "train",</br>    "source": "out = self.cls_net(data_dict['img'])"</br>  },</br>  {</br>    "filename": "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py",</br>    "linenum": 550,</br>    "funcname": "__call__",</br>    "source": "result = self.forward(*input, **kwargs)"</br>  },</br>  {</br>    "filename": "/usr/local/lib/python3.6/dist-packages/torch/nn/parallel/data_parallel.py",</br>    "linenum": 150,</br>    "funcname": "forward",</br>    "source": "return self.module(*inputs[0], **kwargs[0])"</br>  },</br>  {</br>    "filename": "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py",</br>    "linenum": 550,</br>    "funcname": "__call__",</br>    "source": "result = self.forward(*input, **kwargs)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/model/cls/nets/base_model.py",</br>    "linenum": 61,</br>    "funcname": "forward",</br>    "source": "return self.net(inputs)"</br>  },</br>  {</br>    "filename": "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py",</br>    "linenum": 550,</br>    "funcname": "__call__",</br>    "source": "result = self.forward(*input, **kwargs)"</br>  },</br>  {</br>    "filename": "/cache/output/custom/custom_base.py",</br>    "linenum": 19,</br>    "funcname": "forward",</br>    "source": "Flatten_m7okrjm4d_feat = self.Flatten_m7okrjm4d(Input_oth1flre4i_feat)"</br>  }</br>]
100401|size mismatch, m1: [21504 x 224], m2: [10 x 22] at /pytorch/aten/src/THC/generic/THCTensorMathBlas.cu:283|[</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/lib/runner/controller.py",</br>    "linenum": 37,</br>    "funcname": "train",</br>    "source": "runner.train()"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/runner/cls/image_classifier.py",</br>    "linenum": 121,</br>    "funcname": "train",</br>    "source": "out = self.cls_net(data_dict['img'])"</br>  },</br>  {</br>    "filename": "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py",</br>    "linenum": 550,</br>    "funcname": "__call__",</br>    "source": "result = self.forward(*input, **kwargs)"</br>  },</br>  {</br>    "filename": "/usr/local/lib/python3.6/dist-packages/torch/nn/parallel/data_parallel.py",</br>    "linenum": 150,</br>    "funcname": "forward",</br>    "source": "return self.module(*inputs[0], **kwargs[0])"</br>  },</br>  {</br>    "filename": "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py",</br>    "linenum": 550,</br>    "funcname": "__call__",</br>    "source": "result = self.forward(*input, **kwargs)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/model/cls/nets/base_model.py",</br>    "linenum": 61,</br>    "funcname": "forward",</br>    "source": "return self.net(inputs)"</br>  },</br>  {</br>    "filename": "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py",</br>    "linenum": 550,</br>    "funcname": "__call__",</br>    "source": "result = self.forward(*input, **kwargs)"</br>  },</br>  {</br>    "filename": "/cache/output/custom/custom_base.py",</br>    "linenum": 18,</br>    "funcname": "forward",</br>    "source": "Linear_uhzliwikfn_feat = self.Linear_uhzliwikfn(x)"</br>  },</br>  {</br>    "filename": "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py",</br>    "linenum": 550,</br>    "funcname": "__call__",</br>    "source": "result = self.forward(*input, **kwargs)"</br>  },</br>  {</br>    "filename": "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/linear.py",</br>    "linenum": 87,</br>    "funcname": "forward",</br>    "source": "return F.linear(input, self.weight, self.bias)"</br>  },</br>  {</br>    "filename": "/usr/local/lib/python3.6/dist-packages/torch/nn/functional.py",</br>    "linenum": 1612,</br>    "funcname": "linear",</br>    "source": "output = input.matmul(weight.t())"</br>  }</br>]
100403|__init__() missing 1 required positional argument: 'in_channels'|[</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/runner/runner_selector.py",</br>    "linenum": 114,</br>    "funcname": "cls_runner",</br>    "source": "return CLS_METHOD_DICT[key](self.configer)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/runner/cls/image_classifier.py",</br>    "linenum": 39,</br>    "funcname": "__init__",</br>    "source": "self.cls_net = self.cls_model_manager.get_cls_model()"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/model/cls/model_manager.py",</br>    "linenum": 33,</br>    "funcname": "get_cls_model",</br>    "source": "model = CLS_MODEL_DICT[model_name](self.configer)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/app/k12ai/model/cls/nets/custom_base.py",</br>    "linenum": 21,</br>    "funcname": "__init__",</br>    "source": "self.net = load_custom_model(cache_dir, model_name)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/app/k12ai/tools/util/net_def.py",</br>    "linenum": 32,</br>    "funcname": "load_custom_model",</br>    "source": "custom_model = custom_net.custom_model()"</br>  },</br>  {</br>    "filename": "/cache/output/custom/custom_base.py",</br>    "linenum": 29,</br>    "funcname": "custom_model",</br>    "source": "return Hproject_model_name()"</br>  },</br>  {</br>    "filename": "/cache/output/custom/custom_base.py",</br>    "linenum": 11,</br>    "funcname": "__init__",</br>    "source": "self.Conv2d_ptvet7fj4 = nn.Conv2d(out_channels=64, kernel_size=3, stride=1, padding=1, dilation=1, groups=1, bias=True)"</br>  }</br>]


##### Unkown错误

In [61]:
Markdown(unkown_error_md(errors))

ErrorCode | ErrorText | Trackback
    :---:|:---|:---
    <img width=150/> | <img width=250/> | <img width=350/>
    4788bf|__init__() missing 1 required positional argument: 'in_features'|[</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/runner/runner_selector.py",</br>    "linenum": 114,</br>    "funcname": "cls_runner",</br>    "source": "return CLS_METHOD_DICT[key](self.configer)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/runner/cls/image_classifier.py",</br>    "linenum": 39,</br>    "funcname": "__init__",</br>    "source": "self.cls_net = self.cls_model_manager.get_cls_model()"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/model/cls/model_manager.py",</br>    "linenum": 33,</br>    "funcname": "get_cls_model",</br>    "source": "model = CLS_MODEL_DICT[model_name](self.configer)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/app/k12ai/model/cls/nets/custom_base.py",</br>    "linenum": 21,</br>    "funcname": "__init__",</br>    "source": "self.net = load_custom_model(cache_dir, model_name)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/app/k12ai/tools/util/net_def.py",</br>    "linenum": 32,</br>    "funcname": "load_custom_model",</br>    "source": "custom_model = custom_net.custom_model()"</br>  },</br>  {</br>    "filename": "/cache/output/custom/custom_base.py",</br>    "linenum": 27,</br>    "funcname": "custom_model",</br>    "source": "return Hproject_model_name()"</br>  },</br>  {</br>    "filename": "/cache/output/custom/custom_base.py",</br>    "linenum": 15,</br>    "funcname": "__init__",</br>    "source": "self.Linear_x1sub019a = nn.Linear(out_features=22, bias=True)"</br>  }</br>]
173d43|Trying to create tensor with negative dimension -392: [10, -392]|[</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/runner/runner_selector.py",</br>    "linenum": 114,</br>    "funcname": "cls_runner",</br>    "source": "return CLS_METHOD_DICT[key](self.configer)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/runner/cls/image_classifier.py",</br>    "linenum": 39,</br>    "funcname": "__init__",</br>    "source": "self.cls_net = self.cls_model_manager.get_cls_model()"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/model/cls/model_manager.py",</br>    "linenum": 33,</br>    "funcname": "get_cls_model",</br>    "source": "model = CLS_MODEL_DICT[model_name](self.configer)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/app/k12ai/model/cls/nets/custom_base.py",</br>    "linenum": 21,</br>    "funcname": "__init__",</br>    "source": "self.net = load_custom_model(cache_dir, model_name)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/app/k12ai/tools/util/net_def.py",</br>    "linenum": 32,</br>    "funcname": "load_custom_model",</br>    "source": "custom_model = custom_net.custom_model()"</br>  },</br>  {</br>    "filename": "/cache/output/custom/custom_base.py",</br>    "linenum": 37,</br>    "funcname": "custom_model",</br>    "source": "return Hproject_model_name()"</br>  },</br>  {</br>    "filename": "/cache/output/custom/custom_base.py",</br>    "linenum": 19,</br>    "funcname": "__init__",</br>    "source": "self.Linear_dmz8bluxar = nn.Linear(in_features=-392, out_features=10, bias=True)"</br>  },</br>  {</br>    "filename": "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/linear.py",</br>    "linenum": 72,</br>    "funcname": "__init__",</br>    "source": "self.weight = Parameter(torch.Tensor(out_features, in_features))"</br>  }</br>]
dbe4e4|Trying to create tensor with negative dimension -72: [10, -72]|[</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/runner/runner_selector.py",</br>    "linenum": 114,</br>    "funcname": "cls_runner",</br>    "source": "return CLS_METHOD_DICT[key](self.configer)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/runner/cls/image_classifier.py",</br>    "linenum": 39,</br>    "funcname": "__init__",</br>    "source": "self.cls_net = self.cls_model_manager.get_cls_model()"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/model/cls/model_manager.py",</br>    "linenum": 33,</br>    "funcname": "get_cls_model",</br>    "source": "model = CLS_MODEL_DICT[model_name](self.configer)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/app/k12ai/model/cls/nets/custom_base.py",</br>    "linenum": 21,</br>    "funcname": "__init__",</br>    "source": "self.net = load_custom_model(cache_dir, model_name)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/app/k12ai/tools/util/net_def.py",</br>    "linenum": 32,</br>    "funcname": "load_custom_model",</br>    "source": "custom_model = custom_net.custom_model()"</br>  },</br>  {</br>    "filename": "/cache/output/custom/custom_base.py",</br>    "linenum": 37,</br>    "funcname": "custom_model",</br>    "source": "return Hproject_model_name()"</br>  },</br>  {</br>    "filename": "/cache/output/custom/custom_base.py",</br>    "linenum": 19,</br>    "funcname": "__init__",</br>    "source": "self.Linear_dmz8bluxar = nn.Linear(in_features=-72, out_features=10, bias=True)"</br>  },</br>  {</br>    "filename": "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/linear.py",</br>    "linenum": 72,</br>    "funcname": "__init__",</br>    "source": "self.weight = Parameter(torch.Tensor(out_features, in_features))"</br>  }</br>]
723c6b|__init__() missing 1 required positional argument: 'num_features'|[</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/runner/runner_selector.py",</br>    "linenum": 114,</br>    "funcname": "cls_runner",</br>    "source": "return CLS_METHOD_DICT[key](self.configer)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/runner/cls/image_classifier.py",</br>    "linenum": 39,</br>    "funcname": "__init__",</br>    "source": "self.cls_net = self.cls_model_manager.get_cls_model()"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/torchcv/model/cls/model_manager.py",</br>    "linenum": 33,</br>    "funcname": "get_cls_model",</br>    "source": "model = CLS_MODEL_DICT[model_name](self.configer)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/app/k12ai/model/cls/nets/custom_base.py",</br>    "linenum": 21,</br>    "funcname": "__init__",</br>    "source": "self.net = load_custom_model(cache_dir, model_name)"</br>  },</br>  {</br>    "filename": "/hzcsk12/cv/app/k12ai/tools/util/net_def.py",</br>    "linenum": 32,</br>    "funcname": "load_custom_model",</br>    "source": "custom_model = custom_net.custom_model()"</br>  },</br>  {</br>    "filename": "/cache/output/custom/custom_base.py",</br>    "linenum": 33,</br>    "funcname": "custom_model",</br>    "source": "return Hproject_model_name()"</br>  },</br>  {</br>    "filename": "/cache/output/custom/custom_base.py",</br>    "linenum": 15,</br>    "funcname": "__init__",</br>    "source": "self.BatchNorm2d_f8s67obtuj = nn.BatchNorm2d(momentum=0.1, affine=True, track_running_stats=True)"</br>  }</br>]


#### 用户错误分类

In [ ]:
df = pd.read_excel(xls_file, header=0, usecols='A:C')[df.error_code > 100100]

In [62]:
df_grouped = df.groupby(['error_code', 'user_code'], as_index=True)
df_grouped_count = df_grouped.count()
df_grouped_count.columns = ['count']
df_grouped_count.index = df_grouped_count.index.map(lambda x: (f'{x[0]}:{ERRORS[x[0]]["cn"]}', x[1]))
df_grouped_count

count
error_code                         user_code           
100208:任务启动错误: 评估服务缺少模型文件          1011020702001      1
                                   1011020702012      2
100401:自定义模型错误: 模型上下层size大小不匹配     1011020702001      2
100403:自定义模型错误: 卷积层缺少in_channels参数 1011020702024      4
100404:自定义模型错误: 连线错误(箭头顺序)         1011020702025      1
                                   1011020702030      3
999999:常见错误: 不可知错误                 1011020702001      6
                                   1011020702012      2

In [41]:
df_sorted = df.set_index(['error_code', 'user_code']).sort_index()
df_sorted.index = df_sorted.index.map(lambda x: (f'{x[0]}:{ERRORS[x[0]]["cn"]}', x[1]))
df_sorted

mission_id
error_code                         user_code                
100208:任务启动错误: 评估服务缺少模型文件          1011020702001           4
                                   1011020702012           7
                                   1011020702012           7
100401:自定义模型错误: 模型上下层size大小不匹配     1011020702001           4
                                   1011020702001           4
100403:自定义模型错误: 卷积层缺少in_channels参数 1011020702024           7
                                   1011020702024           7
                                   1011020702024           7
                                   1011020702024           7
100404:自定义模型错误: 连线错误(箭头顺序)         1011020702025           7
                                   1011020702030           2
                                   1011020702030           2
                                   1011020702030           2
999999:常见错误: 不可知错误                 1011020702001           4
                                   1011020702001           4
                                   1011020702001           4
                                   1011020702001           4
                                   1011020702001           4
                                   1011020702001           4
                                   1011020702012           7
                                   1011020702012           7